include/grpc/census.h

```diff
@@ -0,0 +1,91 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+/* RPC-internal Census API's. These are designed to be generic enough that
+ * they can (ultimately) be used in many different RPC systems (with differing
+ * implementations). */
+
+#ifndef CENSUS_CENSUS_H
+#define CENSUS_CENSUS_H
+
+#include <grpc/grpc.h>
+#include <grpc/support/slice_buffer.h>
+
+/* Identify census functionality that can be enabled via census_initialize(). */
+enum census_functions {
+  CENSUS_NONE = 0,    /* Do not enable census. */
+  CENSUS_TRACING = 1, /* Enable census tracing. */
+  CENSUS_STATS = 2,   /* Enable Census stats collection. */
+  CENSUS_CPU = 4,     /* Enable Census CPU usage collection. */
+  CENSUS_ALL = CENSUS_TRACING | CENSUS_STATS | CENSUS_CPU
+};
+
+/* Shutdown and startup census subsystem. The 'functions' argument should be
+ * the OR (|) of census_functions values. If census fails to initialize, then
+ * census_initialize() will return a non-zero value. */
+int census_initialize(int functions);
+void census_shutdown();
+
+/* Internally, Census relies on a context, which should be propagated across
+ * RPC's. From the RPC subsystems viewpoint, this is an opaque data structure.
+ * A context must be used as the first argument to all other census
+ * functions. The context can be serialized for passing across the wire. */
+typedef struct census_context census_context;
+
+/* This function is called by the RPC subsystem whenever it needs to get a
+ * serialized form of the current census context (presumably to pass across
+ * the wire). Arguments:
+ * 'buffer': pointer to memory into which serialized context will be placed
+ * 'buf_size': size of 'buffer'
+ *
+ * Returns: the number of bytes used in buffer if successful, or 0 if the
+ * buffer is of insufficient size.
+ *
+ * TODO(aveitch): determine how best to communicate required/max buffer size
+ * so caller doesn't have to guess. */
+size_t census_context_serialize(const census_context *context, char *buffer,
+                                size_t buf_size);
+
+/* Create a new census context, possibly from a serialized buffer. If 'buffer'
+ * is non-NULL, it is assumed that it is a buffer encoded by
+ * census_context_serialize(). If `buffer` is NULL, a new, empty context is
+ * created.
+ *
+ * Returns NULL on error (buffer is incorrectly formatted) */
```

Comment:
Do we want to have predefined unknown_context and default_context?
If we have, maybe we can just return unknown or default context here instead of NULL on error?